File: inverter_hil/tests/inverter_hil/TestSensorConcurrency.m
classdef TestSensorConcurrency < matlab.unittest.TestCase
    methods (Test)
        function SharedStateProducesIndependentFrames(testCase)
            addpath(fullfile(fileparts(fileparts(fileparts(mfilename('fullpath')))), ...
                'imu'));
            addpath(fullfile(fileparts(fileparts(fileparts(mfilename('fullpath')))), ...
                'steering-sensor'));
            vehicle = struct('accelerationMps2', [0.5 -0.25 9.8], ...
                'rateOfTurnRadPerS', [0.1 0.2 0.3]);
            [~, imu] = stepImuSimulation([], 0.010, vehicle, []);
            steeringInput = struct('angleDeg', 30, 'dropout', false, ...
                'status', struct('trim', true, 'cal', true, 'ok', true));
            [~, lws] = stepSteeringSensor([], 0.010, steeringInput);
            frames = inverterhil.sensorBusMux(imu, lws);
            testCase.verifyEqual(numel(frames), 2);
            testCase.verifyEqual([frames.timestampS], [0.010 0.010]);
            testCase.verifyEqual({frames.kind}, {'imu', 'steering'});
        end

        function OneDropoutDoesNotSuppressOtherSensor(testCase)
            addpath(fullfile(fileparts(fileparts(fileparts(mfilename('fullpath')))), ...
                'imu'));
            addpath(fullfile(fileparts(fileparts(fileparts(mfilename('fullpath')))), ...
                'steering-sensor'));
            vehicle = struct('accelerationMps2', [0 0 9.8], ...
                'rateOfTurnRadPerS', [0 0 0]);
            [~, imu] = stepImuSimulation([], 0.020, vehicle, ...
                struct('enabled', true, 'dropout', true));
            [~, lws] = stepSteeringSensor([], 0.020, ...
                struct('angleDeg', -20, 'dropout', false, ...
                'status', struct('trim', true, 'cal', true, 'ok', true)));
            frames = inverterhil.sensorBusMux(imu, lws);
            testCase.verifyEqual(numel(frames), 1);
            testCase.verifyEqual(frames.kind, 'steering');
        end

        function BothSensorsShareTimestampOnFirstTick(testCase)
            previous = [];
            input = struct('imuPeriodS', 0.005, 'lwsPeriodS', 0.010);
            [state, due] = inverterhil.sensorScheduler(previous, 0, input);
            testCase.verifyTrue(due.imu);
            testCase.verifyTrue(due.lws);
            testCase.verifyEqual(due.timestampS, 0);
            testCase.verifyEqual(due.imuSequence, uint32(1));
            testCase.verifyEqual(due.lwsSequence, uint32(1));
            testCase.verifyEqual(state.lastImuS, state.lastLwsS);
        end

        function IndependentRatesDoNotBlockEachOther(testCase)
            input = struct('imuPeriodS', 0.005, 'lwsPeriodS', 0.010);
            [state, due] = inverterhil.sensorScheduler([], 0, input);
            testCase.verifyTrue(due.imu && due.lws);
            [state, due] = inverterhil.sensorScheduler(state, 0.005, input);
            testCase.verifyTrue(due.imu);
            testCase.verifyFalse(due.lws);
            [state, due] = inverterhil.sensorScheduler(state, 0.010, input);
            testCase.verifyTrue(due.imu && due.lws);
            testCase.verifyEqual(state.imuSequence, uint32(3));
            testCase.verifyEqual(state.lwsSequence, uint32(2));
        end

        function SensorPeriodsRejectInvalidConfiguration(testCase)
            testCase.verifyError(@() inverterhil.sensorScheduler([], 0, ...
                struct('imuPeriodS', 0, 'lwsPeriodS', 0.01)), ...
                'inverterhil:InvalidSensorPeriod');
            testCase.verifyError(@() inverterhil.sensorScheduler([], -1, ...
                struct('imuPeriodS', 0.005, 'lwsPeriodS', 0.01)), ...
                'inverterhil:InvalidSensorTime');
        end
    end
end

File: VC_HIL/python/tests/test_sensor_concurrency.py
import pytest
from sensors import stepImuSimulation, stepSteeringSensor
from inverterhil import sensorBusMux, sensorScheduler


def test_shared_state_produces_independent_frames():
    vehicle = {"accelerationMps2": [.5, -.25, 9.8], "rateOfTurnRadPerS": [.1, .2, .3]}
    _, imu, _, _ = stepImuSimulation(None, .010, vehicle)
    _, lws = stepSteeringSensor(None, .010, {"angleDeg": 30, "dropout": False,
                                             "status": {"trim": True, "cal": True, "ok": True}})
    frames = sensorBusMux(imu, lws)
    assert [x["timestampS"] for x in frames] == [.010, .010]
    assert [x["kind"] for x in frames] == ["imu", "steering"]


def test_one_dropout_does_not_suppress_other_sensor():
    vehicle = {"accelerationMps2": [0, 0, 9.8], "rateOfTurnRadPerS": [0, 0, 0]}
    _, imu, _, _ = stepImuSimulation(None, .020, vehicle, {"enabled": True, "dropout": True})
    _, lws = stepSteeringSensor(None, .020, {"angleDeg": -20, "dropout": False,
                                             "status": {"trim": True, "cal": True, "ok": True}})
    assert [x["kind"] for x in sensorBusMux(imu, lws)] == ["steering"]


def test_first_tick_shared_timestamp_and_independent_rates():
    config = {"imuPeriodS": .005, "lwsPeriodS": .010}
    state, due = sensorScheduler(None, 0, config)
    assert due["imu"] and due["lws"] and due["imuSequence"] == due["lwsSequence"] == 1
    state, due = sensorScheduler(state, .005, config)
    assert due["imu"] and not due["lws"]
    state, due = sensorScheduler(state, .010, config)
    assert due["imu"] and due["lws"] and state["imuSequence"] == 3 and state["lwsSequence"] == 2


def test_periods_reject_invalid_configuration():
    with pytest.raises(ValueError): sensorScheduler(None, 0, {"imuPeriodS": 0, "lwsPeriodS": .01})
    with pytest.raises(ValueError): sensorScheduler(None, -1, {"imuPeriodS": .005, "lwsPeriodS": .01})
